app/src/main/java/org/kiwix/kiwixmobile/localFileTransfer/ReceiverDeviceAsyncTask.kt

```diff
@@ -0,0 +1,116 @@
+/**
+ * Helper class for the local file sharing module.
+ *
+ * Once the handshake has successfully taken place, this async-task is used to receive files from
+ * the sender device on the FILE_TRANSFER_PORT port. No. of files to be received (and their names)
+ * are learnt beforehand during the handshake.
+ *
+ * A single Task is used for the entire file transfer (the server socket accepts connections as
+ * many times as the no. of files).
+ */
+internal class ReceiverDeviceAsyncTask(private val wifiDirectManager: WifiDirectManager) :
+  AsyncTask<Void?, Int?, Boolean>() {
+  private var incomingFileName: String? = null
+
+  @SuppressWarnings("NestedBlockDepth")
+  override fun doInBackground(vararg p0: Void?): Boolean? {
+    try {
+      ServerSocket(WifiDirectManager.FILE_TRANSFER_PORT).use { serverSocket ->
+        Log.d(TAG, "Server: Socket opened at " + WifiDirectManager.FILE_TRANSFER_PORT)
+        val zimStorageRootPath = wifiDirectManager.zimStorageRootPath
+        val fileItems = wifiDirectManager.getFilesForTransfer()
+        var isTransferErrorFree = true
+        if (BuildConfig.DEBUG) Log.d(TAG, "Expecting " + fileItems.size + " files")
+        var fileItemIndex = 0
+        while (fileItemIndex < fileItems.size && !isCancelled) {
+          incomingFileName = fileItems[fileItemIndex].fileName
+          try {
+            serverSocket.accept().use { client ->
+              if (BuildConfig.DEBUG) {
+                Log.d(TAG, "Sender device connected for " + fileItems[fileItemIndex].fileName)
+              }
+              publishProgress(fileItemIndex, FileItem.FileStatus.SENDING.ordinal)
+              val clientNoteFileLocation = File(zimStorageRootPath + incomingFileName)
+              val dirs = File(clientNoteFileLocation.parent)
+              if (!dirs.exists() && !dirs.mkdirs()) {
+                Log.d(TAG, "ERROR: Required parent directories couldn't be created")
+                isTransferErrorFree = false
+                fileItemIndex++
+              }
+              val fileCreated = clientNoteFileLocation.createNewFile()
+              if (BuildConfig.DEBUG) Log.d(TAG, "File creation: $fileCreated")
+              copyToOutputStream(client.getInputStream(), FileOutputStream(clientNoteFileLocation))
+              publishProgress(fileItemIndex, FileItem.FileStatus.SENT.ordinal)
+            }
+          } catch (e: IOException) {
+            Log.e(TAG, e.message)
+            isTransferErrorFree = false
+            publishProgress(fileItemIndex, FileItem.FileStatus.ERROR.ordinal)
+          }
+          fileItemIndex++
+        }
+        return@doInBackground !isCancelled && isTransferErrorFree
+      }
+    } catch (e: IOException) {
+      Log.e(TAG, e.message)
+      return false // Returned when an error was encountered during transfer
+    }
+  }
+
+  override fun onProgressUpdate(vararg values: Int?) {
+    val fileIndex = values[0]
+    val fileStatus = FileItem.FileStatus.values()[values[1] ?: -1]
```

Comment:
How do I update with this in coroutines?

Reply:
First of all do you understand all the methods of Asynctask and which threads they operate on? That'll be essential in understanding how to adapt the flow

Reply:
yes, do in background works under background thread. and other methods are invoked on UI thread.

Reply:
yes, I think I got it. these codes will run on the UI thread and there should we update now. Is this right?

Reply:
yes so we don't need to have these separate methods with coroutines, lets take an example in a fragment

```
viewLifecycleScope.launch{
  textView.text= "loadingStarted"
  withContext(Dispatchers.IO){
     aLongWebRequest()
 }
 textview.text= "loadFinished"
}
```
this would execute sequentially as you read it.

another way of writing `viewLifecycleScope.launch{` would be `viewLifecycleScope.launch(Dispathers.Main){`

This is a fairly simple way of switching threads in coroutine land

Reply:
looking at this code there is no possible way it can work

```
        senderDeviceAsyncTask = SenderDeviceAsyncTask(this, activity).also {
          it.execute()
        }

  @Override
  protected Boolean doInBackground(FileItem... fileItems) {
```

for the doInBackground to have this array of `fileItems` they would have to be passed in to the `execute` call.

I was going to talk about how this situation might be better modelled by a `Flow` if we were to think about this much like how we handled `sequences` previously. Work on this ticket should really begin after file transfer has been fixed first.

Reply:
>Work on this ticket should really begin after file transfer has been fixed first.

Is this #2392 ?

Reply:
>for the doInBackground to have this array of fileItems they would have to be passed in to the execute call.

It starts on if `isFileSender` is false i.e on receiver side.
```
else {
        callbacks.onFilesForTransferAvailable(filesForTransfer)
        receiverDeviceAsyncTask = ReceiverDeviceAsyncTask(this).also {
          it.execute()
        }
```

Reply:
yup